chainer/functions/noise/zoneout.py

```diff
@@ -0,0 +1,58 @@
+import numpy
+
+from chainer import cuda
+from chainer import function
+from chainer.utils import type_check
+
+
+class Zoneout(function.Function):
+
+    """Zonepout regularization."""
+
+    def __init__(self, zoneout_ratio):
+        self.zoneout_ratio = zoneout_ratio
+
+    def check_type_forward(self, in_types):
+        type_check.expect(in_types.size() == 2)
+
+    def forward(self, inputs):
+        h, x = inputs
+        xp = cuda.get_array_module(*x)
+        if xp == numpy:
+            flag_x = xp.random.rand(*h[0].shape) >= self.zoneout_ratio
+        else:
+            flag_x = (xp.random.rand(*h[0].shape, dtype=numpy.float32) >=
+                      self.zoneout_ratio)
+        flag_h = xp.ones_like(flag_x) - flag_x
+        self.flag_h = flag_h
+        self.flag_x = flag_x
+        return h * self.flag_h + x * self.flag_x,
+
+    def backward(self, inputs, gy):
+        h, x = inputs
+
+        return gy[0] * self.flag_h, gy[0] * self.flag_x,
+
+
+def zoneout(h, x, ratio=.5, train=True):
+    """Instead of dropping out, units zone out and are set to their previous value.
```

Comment:
Although the wording "zone out" is used in the original paper, I think it is better to avoid to use it as it is figurative. How about "units are stochastically retain their previous value"?